File: solutions/typescript/matching-brackets/1/matching-brackets.ts
const expectedOpening = {
  "]": "[",
  "}": "{",
  ")": "(",
}

export function isPaired(input: string): boolean {
  let stack: string[] = [];
  
  for (let c of input.split("")) {
    switch(c){
      case "[":
      case "{":
      case "(":
        stack.push(c);
        break;
      case "]":
      case "}":
      case ")":
        if (stack.pop() != expectedOpening[c]) return false;
    }
  }

  return stack.length === 0;
}

File: solutions/python/matching-brackets/1/matching_brackets.py
closing = {'{': '}', '[': ']', '(': ')'}

def is_paired(input_string: str) -> bool:
    stack = []
    for c in input_string:
        match c:
            case '[' | '(' | '{':
                stack.append(c)
            case ']' | ')' | '}':
                if len(stack) == 0 or c != closing[stack.pop()]:
                    return False
    return len(stack) == 0